File: backend/app/api/v1/routers/stock.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.deps import get_current_user, get_org_context
from app.models.membership import Membership
from app.models.user import User
from app.schemas.stock import (
    AdjustmentCreate,
    ArticleConfigCreate,
    ArticleConfigOut,
    ArticleVariantOut,
    ArticleWithVariantsOut,
    ConsignmentActionCreate,
    ConsignmentSummaryOut,
    DepotCreate,
    DepotOut,
    DepotUpdate,
    MovementCreate,
    MovementOut,
    StockLevelOut,
    ThresholdSet,
    TransferCreate,
    VariantInput,
)
from app.services.organization_service import PermissionDeniedError
from app.services.record_service import RecordService
from app.services.stock_service import InsufficientStockError, StockError, StockService

router = APIRouter(tags=["stock"])


def _error(exc: Exception) -> HTTPException:
    if isinstance(exc, PermissionDeniedError):
        return HTTPException(status.HTTP_403_FORBIDDEN, str(exc))
    if isinstance(exc, InsufficientStockError):
        return HTTPException(status.HTTP_409_CONFLICT, str(exc))
    return HTTPException(status.HTTP_400_BAD_REQUEST, str(exc))


# --- dépôts -----------------------------------------------------------------------


@router.get("/organizations/{organization_id}/depots", response_model=list[DepotOut])
async def list_depots(membership: Membership = Depends(get_org_context), db: AsyncSession = Depends(get_db)) -> list[DepotOut]:
    depots = await StockService(db).list_depots(membership.organization_id)
    return [DepotOut.model_validate(d) for d in depots]


@router.post("/organizations/{organization_id}/depots", response_model=DepotOut, status_code=status.HTTP_201_CREATED)
async def create_depot(
    payload: DepotCreate,
    user: User = Depends(get_current_user),
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
) -> DepotOut:
    service = StockService(db)
    try:
        depot = await service.create_depot(
            organization_id=membership.organization_id, actor=user, actor_membership=membership, payload=payload
        )
    except (PermissionDeniedError, StockError) as exc:
        raise _error(exc) from exc
    return DepotOut.model_validate(depot)


@router.patch("/organizations/{organization_id}/depots/{depot_id}", response_model=DepotOut)
async def update_depot(
    depot_id: uuid.UUID,
    payload: DepotUpdate,
    user: User = Depends(get_current_user),
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
) -> DepotOut:
    service = StockService(db)
    depot = await service.repo.get_depot(depot_id)
    if depot is None or depot.organization_id != membership.organization_id:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Dépôt introuvable.")
    try:
        depot = await service.update_depot(
            organization_id=membership.organization_id, actor=user, actor_membership=membership, depot=depot, payload=payload
        )
    except (PermissionDeniedError, StockError) as exc:
        raise _error(exc) from exc
    return DepotOut.model_validate(depot)


# --- articles / variantes -----------------------------------------------------------


@router.post(
    "/organizations/{organization_id}/records/{record_id}/article-config",
    response_model=ArticleWithVariantsOut,
    status_code=status.HTTP_201_CREATED,
)
async def configure_article(
    record_id: uuid.UUID,
    payload: ArticleConfigCreate,
    user: User = Depends(get_current_user),
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
) -> ArticleWithVariantsOut:
    record = await RecordService(db).get(membership.organization_id, record_id)
    if record is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Fiche introuvable.")
    service = StockService(db)
    try:
        config, variants = await service.configure_article(
            organization_id=membership.organization_id, actor=user, actor_membership=membership, record=record, payload=payload
        )
    except (PermissionDeniedError, StockError) as exc:
        raise _error(exc) from exc
    return ArticleWithVariantsOut(
        config=ArticleConfigOut.model_validate(config), variants=[ArticleVariantOut.model_validate(v) for v in variants]
    )


@router.post(
    "/organizations/{organization_id}/records/{record_id}/variants",
    response_model=ArticleVariantOut,
    status_code=status.HTTP_201_CREATED,
)
async def add_variant(
    record_id: uuid.UUID,
    payload: VariantInput,
    user: User = Depends(get_current_user),
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
) -> ArticleVariantOut:
    record = await RecordService(db).get(membership.organization_id, record_id)
    if record is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Fiche introuvable.")
    service = StockService(db)
    try:
        variant = await service.add_variant(
            organization_id=membership.organization_id, actor=user, actor_membership=membership, record=record, payload=payload
        )
    except (PermissionDeniedError, StockError) as exc:
        raise _error(exc) from exc
    return ArticleVariantOut.model_validate(variant)


@router.put("/organizations/{organization_id}/variants/{variant_id}/threshold", status_code=status.HTTP_204_NO_CONTENT)
async def set_threshold(
    variant_id: uuid.UUID,
    payload: ThresholdSet,
    user: User = Depends(get_current_user),
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
) -> None:
    service = StockService(db)
    variant = await service.repo.get_variant(variant_id)
    if variant is None or variant.organization_id != membership.organization_id:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Variante introuvable.")
    try:
        await service.set_threshold(
            organization_id=membership.organization_id, actor=user, actor_membership=membership, variant=variant, payload=payload
        )
    except (PermissionDeniedError, StockError) as exc:
        raise _error(exc) from exc


# --- mouvements --------------------------------------------------------------------


@router.post(
    "/organizations/{organization_id}/stock/movements/entry",
    response_model=list[MovementOut],
    status_code=status.HTTP_201_CREATED,
)
async def create_entry(
    payload: MovementCreate,
    user: User = Depends(get_current_user),
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
) -> list[MovementOut]:
    service = StockService(db)
    try:
        movements = await service.record_entry(
            organization_id=membership.organization_id, actor=user, actor_membership=membership, payload=payload
        )
    except (PermissionDeniedError, StockError) as exc:
        raise _error(exc) from exc
    return [MovementOut.model_validate(m) for m in movements]


@router.post(
    "/organizations/{organization_id}/stock/movements/exit", response_model=list[MovementOut], status_code=status.HTTP_201_CREATED
)
async def create_exit(
    payload: MovementCreate,
    user: User = Depends(get_current_user),
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
) -> list[MovementOut]:
    service = StockService(db)
    try:
        movements = await service.record_exit(
            organization_id=membership.organization_id, actor=user, actor_membership=membership, payload=payload
        )
    except (PermissionDeniedError, StockError) as exc:
        raise _error(exc) from exc
    return [MovementOut.model_validate(m) for m in movements]


@router.post(
    "/organizations/{organization_id}/stock/movements/adjustment", response_model=MovementOut, status_code=status.HTTP_201_CREATED
)
async def create_adjustment(
    payload: AdjustmentCreate,
    user: User = Depends(get_current_user),
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
) -> MovementOut:
    service = StockService(db)
    try:
        movement = await service.record_adjustment(
            organization_id=membership.organization_id, actor=user, actor_membership=membership, payload=payload
        )
    except (PermissionDeniedError, StockError) as exc:
        raise _error(exc) from exc
    return MovementOut.model_validate(movement)


@router.post(
    "/organizations/{organization_id}/stock/movements/transfer",
    response_model=list[MovementOut],
    status_code=status.HTTP_201_CREATED,
)
async def create_transfer(
    payload: TransferCreate,
    user: User = Depends(get_current_user),
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
) -> list[MovementOut]:
    service = StockService(db)
    try:
        out_movement, in_movement = await service.record_transfer(
            organization_id=membership.organization_id, actor=user, actor_membership=membership, payload=payload
        )
    except (PermissionDeniedError, StockError) as exc:
        raise _error(exc) from exc
    return [MovementOut.model_validate(out_movement), MovementOut.model_validate(in_movement)]


# --- consignation --------------------------------------------------------------------


@router.post("/organizations/{organization_id}/stock/consignment-actions", status_code=status.HTTP_201_CREATED)
async def record_consignment_action(
    payload: ConsignmentActionCreate,
    user: User = Depends(get_current_user),
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
) -> ConsignmentSummaryOut:
    service = StockService(db)
    try:
        await service.record_consignment_action(
            organization_id=membership.organization_id, actor=user, actor_membership=membership, payload=payload
        )
    except (PermissionDeniedError, StockError) as exc:
        raise _error(exc) from exc
    return await service.get_consignment_summary(membership.organization_id, payload.variant_id, payload.depot_id)


# --- lecture ---------------------------------------------------------------------------


@router.get("/organizations/{organization_id}/stock/levels", response_model=list[StockLevelOut])
async def list_stock_levels(
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
    variant_id: uuid.UUID | None = None,
    depot_id: uuid.UUID | None = None,
) -> list[StockLevelOut]:
    levels = await StockService(db).list_stock_levels(membership.organization_id, variant_id=variant_id, depot_id=depot_id)
    return [StockLevelOut.model_validate(level) for level in levels]


@router.get(
    "/organizations/{organization_id}/stock/consignment-summary",
    response_model=ConsignmentSummaryOut,
)
async def consignment_summary(
    variant_id: uuid.UUID,
    depot_id: uuid.UUID,
    membership: Membership = Depends(get_org_context),
    db: AsyncSession = Depends(get_db),
) -> ConsignmentSummaryOut:
    return await StockService(db).get_consignment_summary(membership.organization_id, variant_id, depot_id)

File: frontend/src/lib/api/stock.ts
import { apiRequest } from "./http";
import type {
  AdjustmentCreate,
  ArticleConfigCreate,
  ArticleVariantOut,
  ArticleWithVariantsOut,
  ConsignmentActionCreate,
  ConsignmentSummaryOut,
  DepotCreate,
  DepotOut,
  DepotThresholdOut,
  DepotUpdate,
  MovementCreate,
  MovementListOut,
  MovementOut,
  StockLevelOut,
  StockLotOut,
  ThresholdSet,
  TransferCreate,
  VariantInput,
} from "./types";

/**
 * Client du module Stock (cahier des charges §7). Même pattern que
 * `records.ts`/`model-definitions.ts` : `accessToken` en premier paramètre,
 * `apiRequest` lève `ApiError` sur tout échec (jamais un succès simulé).
 * Schémas backend : `backend/app/schemas/stock.py`. Route : `backend/app/api/v1/routers/stock.py`.
 *
 * Un article de stock est une fiche (`Record`, nature `stock_item`) — ses
 * données de stock (config, variantes, niveaux, lots, mouvements) sont
 * séparées de `Record.data` et vivent ici, jamais dans `lib/api/records.ts`.
 */

const orgBase = (organizationId: string) => `/organizations/${organizationId}`;

// --- dépôts -----------------------------------------------------------------------

export function listDepots(accessToken: string, organizationId: string): Promise<DepotOut[]> {
  return apiRequest<DepotOut[]>(`${orgBase(organizationId)}/depots`, { accessToken });
}

export function createDepot(
  accessToken: string,
  organizationId: string,
  payload: DepotCreate,
): Promise<DepotOut> {
  return apiRequest<DepotOut>(`${orgBase(organizationId)}/depots`, {
    accessToken,
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updateDepot(
  accessToken: string,
  organizationId: string,
  depotId: string,
  payload: DepotUpdate,
): Promise<DepotOut> {
  return apiRequest<DepotOut>(`${orgBase(organizationId)}/depots/${depotId}`, {
    accessToken,
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

// --- articles / variantes -----------------------------------------------------------

/** 404 (`ApiError.status === 404`) = cette fiche n'est pas encore configurée
 * comme article de stock — état normal juste après création d'une fiche de
 * modèle `stock_item`, pas une erreur à afficher telle quelle (voir `StockPanel`). */
export function getArticle(
  accessToken: string,
  organizationId: string,
  recordId: string,
): Promise<ArticleWithVariantsOut> {
  return apiRequest<ArticleWithVariantsOut>(`${orgBase(organizationId)}/records/${recordId}/article`, {
    accessToken,
  });
}

/** Configure l'article une première fois (unité, prix, déclinaison en
 * variantes, suivi de lots, consignation). Ne peut être appelé qu'une fois par
 * fiche — le backend renvoie déjà l'article existant via `getArticle` sinon. */
export function configureArticle(
  accessToken: string,
  organizationId: string,
  recordId: string,
  payload: ArticleConfigCreate,
): Promise<ArticleWithVariantsOut> {
  return apiRequest<ArticleWithVariantsOut>(`${orgBase(organizationId)}/records/${recordId}/article-config`, {
    accessToken,
    method: "POST",
    body: JSON.stringify(payload),
  });
}

/** Ajoute une variante supplémentaire à un article déjà configuré. */
export function addVariant(
  accessToken: string,
  organizationId: string,
  recordId: string,
  payload: VariantInput,
): Promise<ArticleVariantOut> {
  return apiRequest<ArticleVariantOut>(`${orgBase(organizationId)}/records/${recordId}/variants`, {
    accessToken,
    method: "POST",
    body: JSON.stringify(payload),
  });
}

/** `payload.depot_id: null` règle le seuil global de la variante, sinon le
 * seuil spécifique à un dépôt (§7.2 : "réglable globalement et dépôt par dépôt"). */
export function setVariantThreshold(
  accessToken: string,
  organizationId: string,
  variantId: string,
  payload: ThresholdSet,
): Promise<void> {
  return apiRequest<void>(`${orgBase(organizationId)}/variants/${variantId}/threshold`, {
    accessToken,
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

export function listVariantThresholds(
  accessToken: string,
  organizationId: string,
  variantId: string,
): Promise<DepotThresholdOut[]> {
  return apiRequest<DepotThresholdOut[]>(`${orgBase(organizationId)}/variants/${variantId}/thresholds`, {
    accessToken,
  });
}

// --- mouvements (immuables, additifs — jamais modifiés ni supprimés, §7.3) ---------

export function createEntryMovement(
  accessToken: string,
  organizationId: string,
  payload: MovementCreate,
): Promise<MovementOut[]> {
  return apiRequest<MovementOut[]>(`${orgBase(organizationId)}/stock/movements/entry`, {
    accessToken,
    method: "POST",
    body: JSON.stringify(payload),
  });
}

/** Si l'article suit des lots et qu'aucun `lot_number` n'est fourni, le
 * backend consomme automatiquement au plus ancien (FIFO) — peut alors produire
 * plusieurs mouvements, un par lot entamé. */
export function createExitMovement(
  accessToken: string,
  organizationId: string,
  payload: MovementCreate,
): Promise<MovementOut[]> {
  return apiRequest<MovementOut[]>(`${orgBase(organizationId)}/stock/movements/exit`, {
    accessToken,
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function createAdjustmentMovement(
  accessToken: string,
  organizationId: string,
  payload: AdjustmentCreate,
): Promise<MovementOut> {
  return apiRequest<MovementOut>(`${orgBase(organizationId)}/stock/movements/adjustment`, {
    accessToken,
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function createTransferMovement(
  accessToken: string,
  organizationId: string,
  payload: TransferCreate,
): Promise<MovementOut[]> {
  return apiRequest<MovementOut[]>(`${orgBase(organizationId)}/stock/movements/transfer`, {
    accessToken,
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export interface ListMovementsParams {
  variantId?: string;
  depotId?: string;
  recordId?: string;
  limit: number;
  offset: number;
}

export function listMovements(
  accessToken: string,
  organizationId: string,
  params: ListMovementsParams,
): Promise<MovementListOut> {
  const search = new URLSearchParams();
  if (params.variantId) search.set("variant_id", params.variantId);
  if (params.depotId) search.set("depot_id", params.depotId);
  if (params.recordId) search.set("record_id", params.recordId);
  search.set("limit", String(params.limit));
  search.set("offset", String(params.offset));
  return apiRequest<MovementListOut>(`${orgBase(organizationId)}/stock/movements?${search.toString()}`, {
    accessToken,
  });
}

// --- lecture (niveaux / lots / consignation) ----------------------------------------

export interface ListStockLevelsParams {
  variantId?: string;
  depotId?: string;
}

export function listStockLevels(
  accessToken: string,
  organizationId: string,
  params: ListStockLevelsParams = {},
): Promise<StockLevelOut[]> {
  const search = new URLSearchParams();
  if (params.variantId) search.set("variant_id", params.variantId);
  if (params.depotId) search.set("depot_id", params.depotId);
  const qs = search.toString();
  return apiRequest<StockLevelOut[]>(`${orgBase(organizationId)}/stock/levels${qs ? `?${qs}` : ""}`, {
    accessToken,
  });
}

export interface ListStockLotsParams {
  variantId?: string;
  depotId?: string;
  includeEmpty?: boolean;
  /** AAAA-MM-JJ — ne renvoie que les lots expirant avant cette date. */
  expiringBefore?: string;
}

/** Exclut les lots épuisés par défaut (`includeEmpty: false`). */
export function listStockLots(
  accessToken: string,
  organizationId: string,
  params: ListStockLotsParams = {},
): Promise<StockLotOut[]> {
  const search = new URLSearchParams();
  if (params.variantId) search.set("variant_id", params.variantId);
  if (params.depotId) search.set("depot_id", params.depotId);
  if (params.includeEmpty) search.set("include_empty", "true");
  if (params.expiringBefore) search.set("expiring_before", params.expiringBefore);
  const qs = search.toString();
  return apiRequest<StockLotOut[]>(`${orgBase(organizationId)}/stock/lots${qs ? `?${qs}` : ""}`, {
    accessToken,
  });
}

/** N'a de sens que pour un article dont la config porte `is_consigned: true`
 * (§7.6). `deliver_full` incrémente la circulation (et sort du stock plein) ;
 * `return_empty` la décrémente. */
export function recordConsignmentAction(
  accessToken: string,
  organizationId: string,
  payload: ConsignmentActionCreate,
): Promise<ConsignmentSummaryOut> {
  return apiRequest<ConsignmentSummaryOut>(`${orgBase(organizationId)}/stock/consignment-actions`, {
    accessToken,
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function getConsignmentSummary(
  accessToken: string,
  organizationId: string,
  variantId: string,
  depotId: string,
): Promise<ConsignmentSummaryOut> {
  const search = new URLSearchParams({ variant_id: variantId, depot_id: depotId });
  return apiRequest<ConsignmentSummaryOut>(`${orgBase(organizationId)}/stock/consignment-summary?${search.toString()}`, {
    accessToken,
  });
}
